envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterPermissions.kt

```diff
@@ -0,0 +1,73 @@
+package pl.allegro.tech.servicemesh.envoycontrol.snapshot.resource.listeners.filters
+
+import io.envoyproxy.envoy.api.v2.route.HeaderMatcher
+import io.envoyproxy.envoy.config.rbac.v2.Permission
+import io.envoyproxy.envoy.type.matcher.PathMatcher
+import io.envoyproxy.envoy.type.matcher.StringMatcher
+import pl.allegro.tech.servicemesh.envoycontrol.groups.IncomingEndpoint
+import pl.allegro.tech.servicemesh.envoycontrol.groups.PathMatchingType
+
+class RBACFilterPermissions {
+    fun createCombinedPermissions(incomingEndpoint: IncomingEndpoint): Permission.Builder {
+        val combinedPermissions = Permission.newBuilder()
+        val pathPermission = createPathPermissionForEndpoint(incomingEndpoint)
+
+        if (incomingEndpoint.methods.isNotEmpty()) {
+            val methodPermissions = createMethodPermissions(incomingEndpoint)
+            combinedPermissions.setAndRules(
+                    Permission.Set.newBuilder().addAllRules(listOf(
+                            pathPermission.build(),
+                            methodPermissions.build()
+                    ))
+            )
+        } else {
+            combinedPermissions.setAndRules(
+                    Permission.Set.newBuilder().addAllRules(listOf(pathPermission.build()))
+            )
+        }
+
+        return combinedPermissions
+    }
+
+    fun createPathPermission(path: String, matchingType: PathMatchingType): Permission.Builder {
+        return Permission.newBuilder().setUrlPath(createPathMatcher(path, matchingType))
+    }
+
+    private fun createPathPermissionForEndpoint(incomingEndpoint: IncomingEndpoint): Permission.Builder {
+        return createPathPermission(incomingEndpoint.path, incomingEndpoint.pathMatchingType)
+    }
+
+    private fun createMethodPermissions(incomingEndpoint: IncomingEndpoint): Permission.Builder {
+        val methodPermissions = Permission.newBuilder()
+        val methodPermissionSet = Permission.Set.newBuilder()
+        val methodPermissionsList = incomingEndpoint.methods.map(this::mapMethodToHeaderMatcher)
+        methodPermissionSet.addAllRules(methodPermissionsList)
+        methodPermissions.setOrRules(methodPermissionSet.build())
+
+        return methodPermissions
+    }
```

Comment:
How about this, looks simpler to me.
```suggestion
    fun createCombinedPermissions(incomingEndpoint: IncomingEndpoint): Permission.Builder {
        val permissions = listOfNotNull(
            createPathPermissionForEndpoint(incomingEndpoint),
            createMethodPermissions(incomingEndpoint)
        )
            .map { it.build() }

        return Permission.newBuilder().setAndRules(
            Permission.Set.newBuilder().addAllRules(permissions)
        )
    }

    fun createPathPermission(path: String, matchingType: PathMatchingType): Permission.Builder {
        return Permission.newBuilder().setUrlPath(createPathMatcher(path, matchingType))
    }

    private fun createPathPermissionForEndpoint(incomingEndpoint: IncomingEndpoint): Permission.Builder {
        return createPathPermission(incomingEndpoint.path, incomingEndpoint.pathMatchingType)
    }

    private fun createMethodPermissions(incomingEndpoint: IncomingEndpoint): Permission.Builder? {
        if (incomingEndpoint.methods.isEmpty()) {
            return null
        }
        val methodPermissions = Permission.newBuilder()
        val methodPermissionSet = Permission.Set.newBuilder()
        val methodPermissionsList = incomingEndpoint.methods.map(this::mapMethodToHeaderMatcher)
        methodPermissionSet.addAllRules(methodPermissionsList)
        methodPermissions.setOrRules(methodPermissionSet.build())

        return methodPermissions
    }
```

Reply:
This one is better 👍

Reply:
It's better, applied.
To be clear - this is a code moved from other class, it is not a code added in this PR.